### DISCUSSION TITLE
Grouping by genre / artist / album

### DISCUSSION THREAD

**Discussion by hmoffatt** [gerbera/gerbera#1497] [Ideas]
I would like to have my music in the hierarchy Genre - Artist - Album. Right now under Genres I see each Genre but then every track is in there. Is it possible to group them this way?

**Comment by KarlStraussberger** [comment]
Did you already take a look at the structured layout in js-mode?

**Reply by hmoffatt** [reply]
No I will try it. Otherwise I need to write my own js and use the js mode right?
I was hoping someone else might have already done this and would have code to share.

**Reply by KarlStraussberger** [reply]
If you don't like the structured layout, it's up to you to share. 😄

**Comment by tommyp-vr4** [comment]
hmoffatt - I wanted the same layout too so I had created one a few years back.  I'm far from a JS expert but the import.js below worked.  A few sidepoints I have to mention. I prefer to have albums show up in sequential order of release, not alphabetic (so I prefix the albums with year).  Also, I use a few ID tags to separate things.  I use the "composer" tag as a bigger than a genre category.  For example, I assign christmas as a composer to segregate the christmas music from the rest.  I also use a "mood" tag too.  So if I am entertaining, I'd normally want to hear "rocking" music without ballads getting in the mix, so I use the "mood" tag to help out.   

Again, I'm far from a js expert, I learned by using example from here and from the old mediatomb forums.  I had to mess around with things to get it to work right.  I'm sure there is a better way to do things.  

I added this to my config.xml that I can't figure out how to make it look nice on this messageboard:

>      library-options section
        <id3>        
        <auxdata>
          <add-data tag="TIT1"/>
          <add-data tag="TCON"/>
          <add-data tag="MOOD"/>
          <add-data tag="TXXX:Mood"/>
        </auxdata>
      </id3>


Here's my import.js
[import.txt](https://github.com/gerbera/gerbera/files/6721373/import.txt)

**Comment by hmoffatt** [comment]
Is the difference between the default and structured modes in js import mode just which js function is called from C++? Is there a reason to prefer to override one or the other?

**Reply by tommyp-vr4** [reply]
hmoffatt - It's a preference for me.  I use the "composer" tag to sort between large different types of music.  Perhaps I chose "composer" because it was a tag that mediatomb recognized (I've been using MT/Gerbera for a long time).  So, for example, I would use the composer tag to separate regular music from Christmas music.  As I get into the tree, sometimes I like to sort by genres, decades, years *or* artists.  I suppose you can modify your config.xml to point to the structured mode that came with Gerbera (you need to purge your database and re-add the items and restart Gerbera too) and see if you like it or see if you want to go for the plunge and create a custom one to your liking.  Gerbera is very nice in that you can set up the whole structure/tree exactly the way you like.  Other media servers aren't as flexible.

**Reply by hmoffatt** [reply]
Thanks @tommyp-vr4 . I am starting to work on my import.js now to set up the layout how I want it. I am just wondering if there is a reason to prefer to edit addAudio or addAudioStructured. From my reading of the source code, the layout configuration setting is only used from import.js where it decides which function to call, so the js could actually ignore the setting altogether.

**Reply by KarlStraussberger** [reply]
We moved most of the stuff to `common.js` so you can easily write your own functions in custom.js (must be set in `config.xml`). `structured` is more sophisticated and harder to ajust while `simple` can be used as a starting point to add things.

**Comment by hmoffatt** [comment]
I finally had a chance to get into this, and it turns out to be pretty easy once you understand the example js code. 

My partner and I have separate collections (with different tastes), so now I have it split by owner at the top level (by looking at the object location), then have the genre/artist/album and other categories within the two collections.

Thanks to the developers, this is a really cool feature!
